api-extractor/src/definitions/ApiItem.ts

```diff
@@ -399,6 +419,150 @@ abstract class ApiItem {
     }
     return false;
   }
+
+  /**
+   * This is called by ApiItems to analyze the types that appear in an expression.  For example,
+   * if a Public API function returns a class that is defined in this package, but not exported,
+   * this is a problem. collectTypeReferences() finds all TypeReference child nodes under the
+   * specified node and analyzes each one.
+   */
+  protected collectTypeReferences(node: ts.Node): void {
+    if (node.kind === ts.SyntaxKind.Block || (node.kind >= 262 && node.kind <= 291)) {
+      // Don't traverse into code blocks or JSDoc items; we only care about the function signature
+      return;
+    }
+
+    if (node.kind === ts.SyntaxKind.TypeReference) {
+      const typeReference: ts.TypeReferenceNode = node as ts.TypeReferenceNode;
+      this._processTypeReference(typeReference);
+    }
+
+    // Recurse the tree
+    for (const childNode of node.getChildren()) {
+      this.collectTypeReferences(childNode);
+    }
+  }
+
+  /**
+   * This is a helper for collectTypeReferences().  It analyzes a single TypeReferenceNode.
+   */
+  private _processTypeReference(typeReferenceNode: ts.TypeReferenceNode): void {
+    const symbol: ts.Symbol = this.extractor.typeChecker.getSymbolAtLocation(typeReferenceNode.typeName);
+    if (!symbol) {
+      // Is this bad?
+      return;
+    }
+
+    if (symbol.flags & ts.SymbolFlags.TypeParameter) {
+      // Don't analyze e.g. "T" in "Set<T>"
+      return;
+    }
+
+    const typeName: string = typeReferenceNode.typeName.getText();
+
+    let currentSymbol: ts.Symbol = symbol;
+    while (true) {
+      const importTypeReference: IImportTypeReference = this._parseImportStatement(currentSymbol);
+      if (importTypeReference) {
+        if (importTypeReference.packageName.substr(0, 2) !== './') {
+          // [CASE 1] External/Resolved
```

Comment:
BTW Step 2 looks a lot like the code you added to Extractor.analyze(), so you should try to share the implementation.